airflow/executors/dask_executor.py

```diff
@@ -66,20 +63,17 @@ def execute_async(self,
                       command: CommandType,
                       queue: Optional[str] = None,
                       executor_config: Optional[Any] = None) -> None:
-        if not self.futures:
-            raise AirflowException("Executor should be started first.")
+        assert self.futures, NOT_STARTED_MESSAGE
```

Comment:
I would probably leave the `if` conditions instead of asserts

Reply:
I am happy to discuss it and see other's opinion there.  I was myself doubtful about asserts but they are really nicer and in many cases they simply make sense (see above) my comments. And if we discourage asserts we should remove them everywhere (they were used in other parts of the code).

Reply:
I think this is a good case for asserts.

Under normal circumstances they should pass, and it's only if you're developing really that you might get this wrong, so having this be asserts that could be disabled by `python -O` seems like the right thing to do.

Reply:
I'd rather check our options first. How we may avoid assertions.

1. If we need a oneliner: `if not self.futures: raise Error(NOT_STARTED_MESSAGE)`. It is not very elegant but easy to understand.

1. Another option to consider is writing a single checker and call it wether we need it:

    ```python
    def _check_started(self):
            if None in (self.client, self.futures):
                raise AirflowError(NOT_STARTED_MESSAGE)
    ```

1. Or go more pythonic and introduce decorator.
    
    ```python
    def check_futures(fn):
        def wrapped(self):
            if not self.futures:
                raise AirflowError(NOT_STARTED_MESSAGE)```
        return wrapped

    @check_futures
    def execute_async()
        <...>
    ```

1. If we expect that some methods need some time to wait for executor start, [tenacity](https://tenacity.readthedocs.io/en/latest) could be useful:
    
    ```python
    from tenacity import retry

    def raise_not_started():
        raise AirflowError(NOT_STARTED_MESSAGE)

    @retry(stop=stop_after_attempt(3))
    def execute_async()
        try:
            <...>
        except Exception:
            raise AirflowError((NOT_STARTED_MESSAGE)
    ```

My vote goes to decorator solution -- either 3 or 4. Anyway, I think assertions could be occasionally useful for internal checks only. Like you checking all good and raise _your_ exception with meaningful message to end user. Otherwise it looks like a dirty hack. Just a simple AssertionError would confuse average pythonist as developer expects (and official docs clearly states that) assertion usage in test environment only. Confusion harm would outweight readability benefit.

Reply:
solution no.3 is not mypy compatibilie. It is not too smart to understand this. Most often, we want to have assertions to make pylint happy.